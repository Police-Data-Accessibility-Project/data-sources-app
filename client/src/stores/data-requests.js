import { defineStore } from 'pinia';

export const useDataRequestsStore = defineStore('data-requests', {
	state: () => ({
		/** Searches performed during session. */
		cache: {},
		/** Previous route visited - useful for determining whether we are incrementing or decrementing pages in data request by id */
		previousDataRequestRoute: null,
	}),
	getters: {
		getDataRequestFromCache: (state) => (key) => {
			if (!(key in state.cache)) {
				return null;
			}
			return state.cache[key];
		},
	},

	actions: {
		clearCache() {
			this.$patch((state) => {
				state.cache = {};
			});
		},
		setDataRequestToCache(key, data, timestamp = new Date().getTime()) {
			this.$patch((state) => {
				state.cache[key] = {
					data,
					timestamp,
				};
			});
		},
		setPreviousDataRequestRoute(route) {
			this.$patch({
				previousDataSourceRoute: route,
			});
		},
	},
});
