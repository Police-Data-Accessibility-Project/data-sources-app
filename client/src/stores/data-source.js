import { defineStore } from 'pinia';

export const useDataSourceStore = defineStore('data-source', {
	state: () => ({
		/** Previous route visited - useful for determining whether we are incrementing or decrementing pages in data source by id */
		previousDataSourceRoute: null,
		/** Cache by ID -- useful for avoiding unnecessary API calls */
		cache: {},
	}),
	persist: {
		storage: sessionStorage,
	},
	getters: {
		getDataSourceFromCache: (state) => (key) => {
			if (!(key in state.cache)) {
				return null;
			}
			return state.cache[key];
		},
	},
	actions: {
		clearCache() {
			this.$patch((state) => {
				state.cache = {};
			});
		},

		setPreviousDataSourceRoute(route) {
			this.$patch({
				previousDataSourceRoute: route,
			});
		},

		setDataSourceToCache(key, data) {
			this.$patch((state) => {
				state.cache[key] = {
					data,
					timestamp: new Date().getTime(),
				};
			});
		},
	},
});
