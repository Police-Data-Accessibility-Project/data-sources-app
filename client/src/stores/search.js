import { defineStore } from 'pinia';

export const useSearchStore = defineStore('search', {
	state: () => ({
		/** Searches performed during session. */
		cache: {},
		/** Needed for `NEXT` / `BACK` functionality in data source id view */
		mostRecentSearchIds: [],
		/** Needed for `NEXT` / `BACK` functionality in data requests id view */
		mostRecentRequestIds: [],
	}),
	persist: {
		storage: sessionStorage,
		pick: ['mostRecentSearchIds'],
	},
	getters: {
		getSearchFromCache: (state) => (key) => {
			if (!(key in state.cache)) {
				return null;
			}
			return state.cache[key];
		},
	},

	actions: {
		clearCache() {
			this.$patch((state) => {
				state.cache = {};
			});
		},

		setMostRecentSearchIds(ids) {
			this.$patch({
				mostRecentSearchIds: ids,
			});
		},

		setMostRecentRequestIds(ids) {
			this.$patch({
				mostRecentRequestIds: ids,
			});
		},

		setSearchToCache(key, data, timestamp = new Date().getTime()) {
			this.$patch((state) => {
				// Use object notation for setting cache data
				state.cache[key] = {
					data,
					timestamp,
				};
			});
		},
	},
});
