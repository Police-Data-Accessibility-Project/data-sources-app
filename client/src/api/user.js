import axios from 'axios';
import { ENDPOINTS } from './constants';
import { useAuthStore } from '@/stores/auth';
import { useUserStore } from '@/stores/user';

const USER_BASE = `${import.meta.env.VITE_VUE_API_BASE_URL}/user`;
const HEADERS = {
	'Content-Type': 'application/json',
};

export async function changePassword(oldPassword, newPassword) {
	const auth = useAuthStore();
	const user = useUserStore();

	await axios.post(
		`${USER_BASE}/${user.id}/${ENDPOINTS.USER.ID.UPDATE_PASSWORD}`,
		{ old_password: oldPassword, new_password: newPassword },
		{
			headers: {
				...HEADERS,
				Authorization: `Bearer ${auth.$state.tokens.accessToken.value}`,
			},
		},
	);

	return await auth.signInWithEmail(user.email, newPassword);
}
