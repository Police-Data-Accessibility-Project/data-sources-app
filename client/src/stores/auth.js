import axios from 'axios';
import { defineStore } from 'pinia';
import parseJwt from '../util/parseJwt';
import router from '../router';
import { useUserStore } from './user';

const HEADERS = {
	headers: { 'Content-Type': 'application/json' },
};
const LOGIN_URL = `${import.meta.env.VITE_VUE_API_BASE_URL}/login`;
const REFRESH_SESSION_URL = `${import.meta.env.VITE_VUE_API_BASE_URL}/refresh-session`;

export const useAuthStore = defineStore('auth', {
	state: () => ({
		userId: null,
		accessToken: {
			value: null,
			expires: Date.now(),
		},
		returnUrl: null,
	}),
	persist: true,
	actions: {
		async login(email, password) {
			const user = useUserStore();

			try {
				const response = await axios.post(
					LOGIN_URL,
					{ email, password },
					HEADERS,
				);

				// Update user store with email
				user.$patch({ email });

				this.parseTokenAndSetData(response);
				if (this.returnUrl) router.push(this.returnUrl);
			} catch (error) {
				throw new Error(error.response?.data?.message);
			}
		},

		logout(isAuthRoute) {
			const user = useUserStore();

			this.$patch({
				userId: null,
				accessToken: { value: null, expires: Date.now() },
				returnUrl: null,
			});

			user.$patch({
				email: '',
			});
			if (isAuthRoute) router.push('/login');
		},

		async refreshAccessToken() {
			if (!this.$state.userId) return;
			try {
				const response = await axios.post(
					REFRESH_SESSION_URL,
					{ session_token: this.$state.accessToken.value },
					HEADERS,
				);
				return this.parseTokenAndSetData(response);
			} catch (error) {
				throw new Error(error.response?.data?.message);
			}
		},

		parseTokenAndSetData(response) {
			const token = response.data.data;
			const tokenParsed = parseJwt(token);

			this.$patch({
				userId: tokenParsed.sub,
				accessToken: {
					value: token,
					expires: new Date(tokenParsed.exp * 1000).getTime(),
				},
			});
		},
	},
});
