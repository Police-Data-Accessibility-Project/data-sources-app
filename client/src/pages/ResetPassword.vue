<template>
	<main v-if="success" class="pdap-flex-container">
		<h1>Success</h1>
		<p data-test="success-subheading">
			{{
				token
					? 'Your password has been successfully updated'
					: 'We sent you an email with a link to reset your password. Please follow the link in the email to proceed'
			}}
		</p>
		<RouterLink v-if="token" to="/login">Click here to log in </RouterLink>
	</main>
	<main
		v-else-if="!success && token"
		class="pdap-flex-container mx-auto max-w-2xl"
	>
		<h1>Change your password</h1>
		<p
			v-if="isExpiredToken"
			data-test="token-expired"
			class="flex flex-col items-start sm:flex-row sm:items-center sm:gap-4"
		>
			Sorry, that token has expired.
			<RouterLink
				data-test="re-request-link"
				to="/reset-password"
				@click="
					isExpiredToken = false;
					error = undefined;
					token = undefined;
				"
			>
				Click here to request another
			</RouterLink>
		</p>

		<Form
			v-else
			id="reset-password"
			data-test="reset-password-form"
			class="flex flex-col"
			name="reset-password"
			:error="error"
			:schema="FORM_SCHEMA_CHANGE_PASSWORD"
			@change="handleChangeOnError"
			@submit="onSubmitChangePassword"
		>
			<Button class="max-w-full" type="submit">
				{{ loading ? 'Loading...' : 'Change password' }}
			</Button>
		</Form>
	</main>

	<main v-else class="pdap-flex-container mx-auto max-w-2xl">
		<h1>Request a link to reset your password</h1>
		<Form
			id="reset-password"
			data-test="reset-password-form"
			class="flex flex-col"
			name="reset-password"
			:error="error"
			:schema="FORM_SCHEMA_REQUEST_PASSWORD"
			@submit="onSubmitRequestReset"
		>
			<Button class="max-w-full" type="submit">
				{{ loading ? 'Loading...' : 'Request password reset' }}
			</Button>
		</Form>
	</main>
</template>

<script setup>
import { Button, Form } from 'pdap-design-system';
import { useUserStore } from '../stores/user';
import { ref, watchEffect } from 'vue';
import { RouterLink, useRoute } from 'vue-router';

// Constants
const FORM_SCHEMA_CHANGE_PASSWORD = [
	{
		'data-test': 'password',
		id: 'password',
		name: 'password',
		label: 'Password',
		type: 'password',
		placeholder: 'Your updated password',
		value: '',
		validators: {
			password: {
				message: 'Please provide your password',
				value: true,
			},
		},
	},
	{
		'data-test': 'confirm-password',
		id: 'confirmPassword',
		name: 'confirmPassword',
		label: 'Confirm Password',
		type: 'password',
		placeholder: 'Confirm your updated password',
		value: '',
		validators: {
			password: {
				message: 'Please confirm your password',
				value: true,
			},
		},
	},
];

const FORM_SCHEMA_REQUEST_PASSWORD = [
	{
		'data-test': 'email',
		id: 'email',
		name: 'email',
		label: 'Email',
		type: 'text',
		placeholder: 'Your email address',
		value: '',
		validators: {
			email: {
				message: 'Please provide your email address',
				value: true,
			},
		},
	},
];

const {
	params: { token },
} = useRoute();

// Stores
const user = useUserStore();

// Reactive vars
const error = ref(undefined);
const isExpiredToken = ref(false);
const loading = ref(false);
const success = ref(false);

// Effects
// Clear error on success
watchEffect(() => {
	if (success.value) error.value = undefined;
});

// Functions
// Handlers
/**
 * Handles clearing pw-match form errors on change if they exist
 */
function handleChangeOnError(formValues) {
	if (error.value && formValues.password !== formValues.confirmPassword) {
		handlePasswordValidation(formValues);
	}
}

/**
 * Validates that passwords match
 * @returns {boolean} `false` if passwords do not match, `true` if they do
 */
function handlePasswordValidation(formValues) {
	if (formValues.password !== formValues.confirmPassword) {
		if (!error.value) {
			error.value = 'Passwords do not match, please try again.';
		}
		return false;
	} else {
		error.value = undefined;
		return true;
	}
}

/**
 * Updates user's password
 */
async function onSubmitChangePassword(formValues) {
	if (!handlePasswordValidation(formValues)) return;

	try {
		loading.value = true;
		const { password } = formValues;
		await user.resetPassword(password, token);

		success.value = true;
	} catch (err) {
		if (err.message === 'The submitted token is invalid') {
			isExpiredToken.value = true;
		}
		error.value = err.message;
	} finally {
		loading.value = false;
	}
}

/**
 * Updates user's password
 */
async function onSubmitRequestReset(formValues) {
	try {
		loading.value = true;
		const { email } = formValues;
		await user.requestPasswordReset(email);
		success.value = true;
	} catch (err) {
		error.value = err.message;
	} finally {
		loading.value = false;
	}
}
</script>
