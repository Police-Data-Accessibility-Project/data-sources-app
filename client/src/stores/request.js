import { useAuthStore } from '@/stores/auth';
import axios from 'axios';
import { defineStore } from 'pinia';

// Constants
const HEADERS_BASE = {
	authorization: `Basic ${import.meta.env.VITE_ADMIN_API_KEY}`,
	'Content-Type': 'application/json',
};
const REQUESTS_URL = `${import.meta.env.VITE_VUE_API_BASE_URL}/data-requests`;

export const useRequestStore = defineStore('request', {
	state: () => ({}),
	// persist: {
	// 	storage: sessionStorage,
	// },
	actions: {
		async createRequest(data) {
			const { tokens } = useAuthStore();
			const response = await axios.post(REQUESTS_URL, data, {
				headers: {
					...HEADERS_BASE,
					authorization: `Bearer ${tokens.accessToken.value}`,
				},
			});

			return response.data;
		},
	},
});
