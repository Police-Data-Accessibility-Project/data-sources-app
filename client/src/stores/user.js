import axios from 'axios';
import { defineStore } from 'pinia';
import { useAuthStore } from './auth';

const HEADERS = {
	headers: { 'Content-Type': 'application/json' },
};
const SIGNUP_URL = `${import.meta.env.VITE_VUE_API_BASE_URL}/user`;
const CHANGE_PASSWORD_URL = `${import.meta.env.VITE_VUE_API_BASE_URL}/user`;
const REQUEST_PASSWORD_RESET_URL = `${import.meta.env.VITE_VUE_API_BASE_URL}/request-reset-password`;
const PASSWORD_RESET_URL = `${import.meta.env.VITE_VUE_API_BASE_URL}/reset-password`;

export const useUserStore = defineStore('user', {
	state: () => ({
		email: '',
	}),
	persist: true,
	actions: {
		async signup(email, password) {
			const auth = useAuthStore();

			try {
				await axios.post(SIGNUP_URL, { email, password }, HEADERS);
				// Update store with email
				this.$patch({ email });
				// Log users in after signup and return that response
				return await auth.login(email, password);
			} catch (error) {
				throw new Error(error.response?.data?.message);
			}
		},

		async changePassword(email, password) {
			const auth = useAuthStore();
			try {
				await axios.put(
					CHANGE_PASSWORD_URL,
					{ email, password },
					{
						headers: {
							...HEADERS.headers,
							Authorization: `Bearer ${auth.accessToken.value}`,
						},
					},
				);
				return await auth.login(email, password);
			} catch (error) {
				throw new Error(error.response?.data?.message);
			}
		},

		async requestPasswordReset(email) {
			try {
				await axios.post(REQUEST_PASSWORD_RESET_URL, { email }, HEADERS);
			} catch (error) {
				throw new Error(error.response?.data?.message);
			}
		},

		async resetPassword(password, token) {
			try {
				await axios.post(`${PASSWORD_RESET_URL}`, { password, token }, HEADERS);
			} catch (error) {
				throw new Error(error.response?.data?.message);
			}
		},
	},
});
