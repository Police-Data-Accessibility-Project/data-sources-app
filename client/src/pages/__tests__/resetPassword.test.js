import { RouterLinkStub, flushPromises, mount } from '@vue/test-utils';
import { describe, expect, beforeEach, it, vi, beforeAll } from 'vitest';
import { nextTick } from 'vue';
import { createTestingPinia } from '@pinia/testing';
import { useUserStore } from '../../stores/user';
import ResetPassword from '../ResetPassword.vue';

import { useRoute } from 'vue-router';

vi.mock('vue-router', async () => ({
	useRoute: vi.fn(),
	createRouter: vi.fn(() => ({
		beforeEach: vi.fn(),
	})),
	createWebHistory: vi.fn(),
	RouterLink: RouterLinkStub,
}));

let wrapper;
let user;

describe('Reset password page', () => {
	beforeEach(() => {
		wrapper = mount(ResetPassword, {
			global: {
				plugins: [createTestingPinia()],
			},
			stubs: {
				RouterLink: RouterLinkStub,
			},
		});

		user = useUserStore();
	});

	describe('No token (request PW reset)', () => {
		beforeAll(() => {
			useRoute.mockImplementation(() => ({
				params: {
					token: undefined,
				},
			}));
		});

		describe('No token, success', () => {
			it('Renders success message for request reset link', async () => {
				wrapper.vm.success = true;
				await nextTick();

				const subheading = await wrapper.find(
					'[data-test="success-subheading"]',
				);

				expect(subheading.text()).toBe(
					'We sent you an email with a link to reset your password. Please follow the link in the email to proceed',
				);

				expect(wrapper.html()).toMatchSnapshot();
			});
		});

		describe('No token, request reset password link', () => {
			let email;
			let form;

			beforeEach(() => {
				email = wrapper.find('[data-test="email"] input');
				form = wrapper.find('[data-test="reset-password-form"]');
			});

			it('Calls the request reset password method with valid data', async () => {
				await email.setValue('hello@hello.com');

				expect(wrapper.html()).toMatchSnapshot();

				form.trigger('submit');
				await flushPromises();

				expect(user.requestPasswordReset).toHaveBeenCalledWith(
					'hello@hello.com',
				);
			});

			it('Handles API error', async () => {
				const mockError = new Error('foo');
				vi.mocked(user.requestPasswordReset).mockRejectedValueOnce(mockError);

				form.trigger('submit');
				await flushPromises();

				const error = form.find('.pdap-form-error-message');
				expect(error.exists()).toBe(true);
				expect(error.text()).toBe('foo');
			});
		});
	});

	describe('With token (reset password)', () => {
		beforeAll(() => {
			useRoute.mockImplementation(() => ({
				params: {
					token: '123abc',
				},
			}));
		});

		describe('With token, success', () => {
			it('Renders success message for reset password', async () => {
				wrapper.vm.success = true;
				await nextTick();

				const subheading = await wrapper.find(
					'[data-test="success-subheading"]',
				);

				expect(subheading.text()).toBe(
					'Your password has been successfully updated',
				);

				expect(wrapper.html()).toMatchSnapshot();
			});
		});

		describe('With token, token expired', () => {
			it('Renders token expired UI', async () => {
				wrapper.vm.isExpiredToken = true;
				await nextTick();

				const expired = await wrapper.find('[data-test="token-expired"]');
				const reRequest = await wrapper.find('[data-test="re-request-link"]');

				expect(expired.exists()).toBe(true);
				expect(reRequest.exists()).toBe(true);
			});
		});

		describe('With token, reset password', () => {
			let password;
			let confirmPassword;
			let form;

			beforeEach(() => {
				password = wrapper.find('[data-test="password"] input');
				confirmPassword = wrapper.find('[data-test="confirm-password"] input');
				form = wrapper.find('[data-test="reset-password-form"]');
			});

			it('Calls the reset password method with valid data', async () => {
				expect(wrapper.html()).toMatchSnapshot();

				await password.setValue('Password1!');
				await confirmPassword.setValue('Password1!');

				form.trigger('submit');
				await flushPromises();

				expect(user.resetPassword).toHaveBeenCalledOnce();
			});

			it('Renders error message with mismatched passwords when trying to sign up and re-validates form', async () => {
				await password.setValue('Password1!');
				await confirmPassword.setValue('Password41234!');

				await form.trigger('submit');
				await nextTick();
				await flushPromises();

				const error = form.find('.pdap-form-error-message');

				expect(error.exists()).toBe(true);
				expect(error.text()).toBe('Passwords do not match, please try again.');

				await nextTick();

				await confirmPassword.setValue('Pasdasdfasdf');
				await nextTick();
				expect(error.exists()).toBe(true);

				expect(wrapper.html()).toMatchSnapshot();
			});

			it('Handles API error with invalid token', async () => {
				const mockError = new Error('The submitted token is invalid');
				vi.mocked(user.resetPassword).mockRejectedValueOnce(mockError);

				form.trigger('submit');
				await flushPromises();

				const expired = await wrapper.find('[data-test="token-expired"]');
				const reRequest = await wrapper.find('[data-test="re-request-link"]');

				expect(expired.exists()).toBe(true);
				expect(reRequest.exists()).toBe(true);
			});
		});
	});
});
