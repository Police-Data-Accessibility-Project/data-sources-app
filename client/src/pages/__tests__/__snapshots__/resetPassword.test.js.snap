// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Reset password page > No token (request PW reset) > No token, request reset password link > Calls the request reset password method with valid data 1`] = `
<main class="pdap-flex-container mx-auto max-w-2xl">
  <h1>Request a link to reset your password</h1>
  <form class="pdap-form flex flex-col" id="reset-password" name="reset-password">
    <!---->
    <div class="pdap-grid-container pdap-input">
      <input id="email" name="email" placeholder="Your email address" type="text">
      <!---->
      <!---->
      <label class="pdap-input-label" for="email">Email</label>
    </div>
    <button class="pdap-button pdap-button-primary max-w-full" type="submit">Request password reset</button>
  </form>
</main>
`;

exports[`Reset password page > No token (request PW reset) > No token, success > Renders success message for request reset link 1`] = `
<main class="pdap-flex-container">
  <h1>Success</h1>
  <p>We sent you an email with a link to reset your password. Please follow the link in the email to proceed</p>
  <!--v-if-->
</main>
`;

exports[`Reset password page > With token (reset password) > With token, reset password > Calls the reset password method with valid data 1`] = `
<main class="pdap-flex-container mx-auto max-w-2xl">
  <h1>Change your password</h1>
  <form class="pdap-form flex flex-col" id="reset-password" name="reset-password">
    <!---->
    <div class="pdap-grid-container pdap-input">
      <!---->
      <input id="password" name="password" placeholder="Your updated password" type="password">
      <!---->
      <label class="pdap-input-label" for="password">Password</label>
    </div>
    <div class="pdap-grid-container pdap-input">
      <!---->
      <input id="confirmPassword" name="confirmPassword" placeholder="Confirm your updated password" type="password">
      <!---->
      <label class="pdap-input-label" for="confirmPassword">Confirm Password</label>
    </div>
    <button class="pdap-button pdap-button-primary max-w-full" type="submit">Change password</button>
  </form>
</main>
`;

exports[`Reset password page > With token (reset password) > With token, reset password > Renders error message with mismatched passwords when trying to sign up and re-validates form 1`] = `
<main class="pdap-flex-container mx-auto max-w-2xl">
  <h1>Change your password</h1>
  <form class="pdap-form flex flex-col" id="reset-password" name="reset-password">
    <div class="pdap-form-error-message">Passwords do not match, please try again.</div>
    <div class="pdap-grid-container pdap-input">
      <!---->
      <input id="password" name="password" placeholder="Your updated password" type="password">
      <!---->
      <label class="pdap-input-label" for="password">Password</label>
    </div>
    <div class="pdap-grid-container pdap-input">
      <!---->
      <input id="confirmPassword" name="confirmPassword" placeholder="Confirm your updated password" type="password">
      <!---->
      <label class="pdap-input-label" for="confirmPassword">Confirm Password</label>
    </div>
    <button class="pdap-button pdap-button-primary max-w-full" type="submit">Change password</button>
  </form>
</main>
`;

exports[`Reset password page > With token (reset password) > With token, success > Renders success message for reset password 1`] = `
<main class="pdap-flex-container">
  <h1>Success</h1>
  <p>Your password has been successfully updated</p>
  <a>Click here to log in </a>
</main>
`;
